Sources/Partial/Codable/PartialCodable.swift

```diff
@@ -0,0 +1,28 @@
+/// A type that can provide a map of key paths and coding keys to enable
+/// `Codable` conformance to `Partial`.
+public protocol PartialCodable {
+    /// The `CodingKey` type used to encode and decode values.
+    ///
+    /// This type has the extra requirement of being `Hashable`. This conformance is
+    /// synthesized when using a `Hashable` type as the `RawValue` of a `CodingKey` `enum`.
+    associatedtype CodingKey: Swift.CodingKey & Hashable
+
+    /// Generate and return a collection of key paths and coding key maps.
+    ///
+    /// The `KeyPathCodingKeyCollectionBuilder` result builder is provided to simplify the
+    /// implementation of this property, for example:
+    ///
+    /// ```swift
+    /// struct CodableType: Codable, PartialCodable, Hashable {
+    ///    @KeyPathCodingKeyCollectionBuilder<Self, CodingKeys>
+    ///    static var keyPathCodingKeyCollection: KeyPathCodingKeyCollection<Self, CodingKeys> {
+    ///        (\Self.stringValue, CodingKeys.stringValue)
+    ///        (\Self.intValue, CodingKeys.intValue)
+    ///    }
```

Comment:
>Would this example code result in a new KeyPathCodingKeyCollection being created every time keyPathCodingKeyCollection is called?

A new collection is created every call, but it's only created once per encode/decode.

> An alternative implementation would be similar...

I initially had it like this but there's an issue knowing what the type to encode/decode is, so the implementer has to do the casting itself.

The best I could come up was https://github.com/JosephDuffy/Partial/blob/8a99ebbc1c6af9cd520da7654bcf327113eaf2c5/Tests/PartialTests/Tests/PartialCodableTests.swift but I wasn't a fan of it.

> The other possible issue with using @resultBuilder here is that it potentially increases the risk of missing a CodingKeys value.

This is a concern for sure, especially since Swift will synthesise `CodingKeys`. Fixing this would require at least 2 functions (from key path to coding key and visa-versa) which for me is an acceptable traidoff, but I couldn't find a way to do this without the implementation also needing to perform a cast to the expected type.